src/hotspot/share/opto/loopTransform.cpp

```diff
@@ -2851,6 +2851,13 @@ int PhaseIdealLoop::do_range_check(IdealLoopTree *loop, Node_List &old_new) {
     register_new_node(main_cmp, main_cle->in(0));
     _igvn.replace_input_of(main_bol, 1, main_cmp);
   }
+  assert(main_limit == cl->limit() || get_ctrl(main_limit) == pre_ctrl, "wrong control for added limit");
+  const TypeInt* orig_limit_t = _igvn.type(orig_limit)->is_int();
+  bool upward = cl->stride_con() > 0;
+  main_limit = new CastIINode(main_limit, TypeInt::make(upward ? min_jint : orig_limit_t->_lo,
+                                                        upward ? orig_limit_t->_hi : max_jint, Type::WidenMax));
```

Comment:
First, need comment.
Second, why you used min_jint and max_jint instead of main_limit->_lo and main_limit->_hi?

Reply:
Thanks for the reviewing this.

> Second, why you used min_jint and max_jint instead of main_limit->_lo and main_limit->_hi?

Range check elimination changes the main loop's limit. For a loop going up, the new limit is less or equal to the initial limit. So it's less than main_limit->_hi. But it's not guaranteed to be greater or equal to main_limit->_lo. Say, main_limit is a constant, then main_limit->_lo = main_limit->_hi. The new limit can't be >= main_limit->_lo otherwise the new limit is main_limit->_lo = main_limit->_hi and the limit is unchanged.

Reply:
> First, need comment.

Comment added.

Reply:
What are you saying is that low limit does not matter when stride > 0 and we can use min_jint instead of complex expression (if we want correct value). Is it right?